teapots/choreographer-30fps/src/main/cpp/CMakeLists.txt

```diff
@@ -17,36 +17,53 @@
 cmake_minimum_required(VERSION 3.4.1)
 
 # build native_app_glue as a static lib
-add_library(native_app_glue STATIC
-    ${ANDROID_NDK}/sources/android/native_app_glue/android_native_app_glue.c)
-
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11 -Wall -fno-exceptions -fno-rtti")
+add_library(NativeAppGlue STATIC
+  ${ANDROID_NDK}/sources/android/native_app_glue/android_native_app_glue.c)
 
 # build the ndk-helper library
-set(ndk_helper_dir ../../../../common/ndk_helper)
-add_subdirectory(${ndk_helper_dir} ndk_helper)
-
-# Export ANativeActivity_onCreate(), 
-# Refer to: https://github.com/android-ndk/ndk/issues/381.
-set(CMAKE_SHARED_LINKER_FLAGS
-    "${CMAKE_SHARED_LINKER_FLAGS} -u ANativeActivity_onCreate")
+set(ndkHelperSrc ../../../../common/ndk_helper)
+add_subdirectory(${ndkHelperSrc} ndkHelperBin)
 
 # now build app's shared lib
-add_library(ChoreographerNativeActivity SHARED
+add_library(ChoreographerNativeActivity
+  SHARED
     ChoreographerNativeActivity.cpp
-    TeapotRenderer.cpp)
+    TeapotRenderer.cpp
+)
+
+set_target_properties(ChoreographerNativeActivity
+  PROPERTIES
+    CXX_STANDARD 11
+    CXX_STANDARD_REQUIRED YES
+    CXX_EXTENSIONS NO
+)
+target_compile_features(ChoreographerNativeActivity
+  PRIVATE cxx_noexcept
+)
+target_compile_options(ChoreographerNativeActivity
+  PRIVATE -Wall -Werror -fno-rtti
+)
 
-target_include_directories(ChoreographerNativeActivity PRIVATE
+# Export ANativeActivity_onCreate(),
+# Refer to: https://github.com/android-ndk/ndk/issues/381.
+set(CMAKE_SHARED_LINKER_FLAGS
+  "${CMAKE_SHARED_LINKER_FLAGS} -u ANativeActivity_onCreate")
+
+target_include_directories(ChoreographerNativeActivity
+  PRIVATE
     ${ANDROID_NDK}/sources/android/cpufeatures
     ${ANDROID_NDK}/sources/android/native_app_glue
-    ${ndk_helper_dir})
+    ${ndkHelperSrc}
+)
 
 # add lib dependencies
 target_link_libraries(ChoreographerNativeActivity
+  PRIVATE
     android
-    native_app_glue
+    log
     atomic
     EGL
     GLESv2
-    log
-    ndk-helper)
+    NativeAppGlue
+    NdkHelper
+)
```

Comment:
Is there any particular reason for changing the ordering here? Link order does actually matter (it needs to be a topological sort of the dependency graph). If NativeAppGlue of NdkHelper use liblog then this is wrong.

Reply:
yeah, the order is: which one closer to the final module so file, get to the last. More basic a lib is. the fore front it is.  we could adopt the opposite sequence: the more closer to the final so file, the more front it is.  Alphabetic order is ok, but more content aware logical way is better.

Same goes for header file include sequence, although we are not reinforce it too much ( framework guys seems to be good at this )

Reply:
More to the point.  Let CMake do your ordering work for you!

That is to say, ndkHelper and NativeAppGlue both might depend upon log.  If both are targets, you can use:
target_link_libraries(ndkHelper log)
target_link_libraries(NativeAppGlue log)
target_link_libraries(ChoreographerNativeActivity nativeAppGlue ndkhelper)

log gets included before either nativeAppGlue and ndkHelper and doesn't get repeated.  The real power of the target like system is to never have to worry about platform ordering.  Cmake will handle it for you.

Reply:
> More basic a lib is. the fore front it is.

That's the opposite of what it needs to be. If this is working then the libraries that are having their order shuffled here don't depend on each other, but if that's the case then why are you changing it?